File: Daily Exercises/Mar 2021/2021-03-23 : 341. Flatten Nested List Iterator/341. Flatten Nested List Iterator.py
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
#class NestedInteger:
#    def isInteger(self) -> bool:
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        """
#
#    def getInteger(self) -> int:
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        """
#
#    def getList(self) -> [NestedInteger]:
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        """

class NestedIterator:
    def __init__(self, nestedList: [NestedInteger]):
        # store the reversed 'nestedList'
        self.s=nestedList[::-1]
    
    def next(self) -> int:
        # 'hasNext' ensures the top is interger
        # and thus return top directly
        return self.s.pop().getInteger()
    
    def hasNext(self) -> bool:
        # if the top element is nestedlist
        # pop it and reverse it and then push it back
        # until the top is integer
        while len(self.s) and not self.s[-1].isInteger():
            self.s+=self.s.pop().getList()[::-1]
        return bool(len(self.s))


# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())

File: Daily Exercises/Mar 2021/2021-03-23 : 341. Flatten Nested List Iterator/341. Flatten Nested List Iterator.java
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * public interface NestedInteger {
 *
 *     // @return true if this NestedInteger holds a single integer, rather than a nested list.
 *     public boolean isInteger();
 *
 *     // @return the single integer that this NestedInteger holds, if it holds a single integer
 *     // Return null if this NestedInteger holds a nested list
 *     public Integer getInteger();
 *
 *     // @return the nested list that this NestedInteger holds, if it holds a nested list
 *     // Return null if this NestedInteger holds a single integer
 *     public List<NestedInteger> getList();
 * }
 */
public class NestedIterator implements Iterator<Integer> {
    private Deque<Iterator<NestedInteger>> stack;

    public NestedIterator(List<NestedInteger> nestedList) {
        stack= new LinkedList<Iterator<NestedInteger>>();
        stack.push(nestedList.iterator());
    }

    // under the premise that every time 'next' is called,
    // 'hasNext' will be called first,
    // return the top element, and move the iterator
    @Override
    public Integer next() {
        return stack.peek().next().getInteger();
        
    }

    @Override
    public boolean hasNext() {
        while (!stack.isEmpty()) {
            Iterator<NestedInteger> it = stack.peek();
            if (!it.hasNext()) { // the last element is found, pop it
                stack.pop();
                continue;
            }
            NestedInteger nest = it.next();
            // if the current element is 'integer', create another nestedlist
            // and push it to the stack
            if (nest.isInteger()) {
                List<NestedInteger> list = new ArrayList<NestedInteger>();
                list.add(nest);
                stack.push(list.iterator());
                return true;
            }
            // if the current element is 'list', push it into stack
            // and move the iterator pointer to it 
            stack.push(nest.getList().iterator());
        }
        return false;
    }
}

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i = new NestedIterator(nestedList);
 * while (i.hasNext()) v[f()] = i.next();
 */
